bitcoin/src/address/error.rs

```diff
@@ -125,3 +92,76 @@ impl fmt::Display for UnknownAddressTypeError {
 }
 
 impl_std_error!(UnknownAddressTypeError);
+
+/// Address parsing error.
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[non_exhaustive]
+pub enum ParseError {
+    /// Base58 error.
+    Base58(base58::Error),
+    /// Bech32 error.
+    Bech32(bech32::Error),
+    /// The bech32 payload was empty.
+    EmptyBech32Payload,
+    /// The wrong checksum algorithm was used. See BIP-0350.
+    InvalidBech32Variant {
+        /// Bech32 variant that is required by the used Witness version.
+        expected: bech32::Variant,
+        /// The actual Bech32 variant encoded in the address representation.
+        found: bech32::Variant,
+    },
+    /// A witness version conversion/parsing error.
+    WitnessVersion(witness_version::Error),
+    /// A witness program error.
+    WitnessProgram(witness_program::Error),
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ParseError::*;
+
+        match *self {
+            Base58(ref e) => write_err!(f, "base58 error"; e),
+            Bech32(ref e) => write_err!(f, "bech32 error"; e),
+            EmptyBech32Payload => write!(f, "the bech32 payload was empty"),
+            InvalidBech32Variant { expected, found } => write!(
+                f,
+                "invalid bech32 checksum variant found {:?} when {:?} was expected",
+                found, expected
+            ),
+            WitnessVersion(ref e) => write_err!(f, "witness version conversion/parsing error"; e),
+            WitnessProgram(ref e) => write_err!(f, "witness program error"; e),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ParseError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use ParseError::*;
+
+        match *self {
+            Base58(ref e) => Some(e),
+            Bech32(ref e) => Some(e),
+            WitnessVersion(ref e) => Some(e),
+            WitnessProgram(ref e) => Some(e),
+            EmptyBech32Payload | InvalidBech32Variant { .. } => None,
+        }
+    }
+}
+
+impl From<base58::Error> for ParseError {
```

Comment:
I don't think it is approriate to create `From<base58::Error>` for `ParseError`. Can we wrap this in a struct `struct AddrBase58Error(base58::Error)`. And then in the enum we would have variant as `Base58(AddrBase58Error)` instead of `Base58(base58::Error)`. 
Finally, we can implement `From<AddrBase58Error> for ParseError`.

Reply:
Thanks for the review. I don't understand your concern, can you elaborate please? Why can't we put the `base58::Error` directly in a variant, this is "standard" error handling boilerplate?

Reply:
Consider the following code. 
```rust
fn foo() -> ParseError {
     let sender_name = base58::decode(...)?; // Some other usage of base58 completely unrelated to address module
     let addr_parsed = Address::from_str(&base58_str)?; // Error related to base58 in address module.
}
```

In the current implementation, the above code would compile both of them will go into `address::ParseError()`. 
But sender name is really not related to address.

Reply:
Fair point but this same applies for every error we have, right? Just to clarify `foo` in your example is a function in downstream code that is returning an error from our code? Or are you saying that we might add `foo` in rust-bitcoin at some later stage and inadvertently create the situation you describe?

Said another way, I don't think the `From` imples are the problem, I think the `foo() -> ParseError`  function is buggy because it returns the wrong error type. Am I understanding you correctly?

Reply:
> Fair point but this same applies for every error we have, right?

Agreed. I no longer have the above concern. Will do a detailed review later.

Reply:
No worries, thanks man.